src/hdf5_interface.F90

```diff
@@ -1551,33 +1637,122 @@ subroutine read_string(group_id, name, buffer, indep)
+    end if
+
+    call h5dclose_f(dset, hdf5_err)
+    call h5sclose_f(dspace, hdf5_err)
+    call h5tclose_f(memtype, hdf5_err)
+    call h5tclose_f(filetype, hdf5_err)
+  end subroutine write_string_1D_explicit
+
+!===============================================================================
+! READ_STRING_1D reads string 1-D array data
+!===============================================================================
 
+  subroutine read_string_1D(group_id, name, buffer, indep)
+    integer(HID_T), intent(in) :: group_id
+    character(*), intent(in)            :: name
+    character(*), intent(inout), target :: buffer(:)
+    logical,      intent(in), optional  :: indep ! independent I/O
+
+    integer(HSIZE_T) :: dims(1)
+
+    dims(:) = shape(buffer)
+    if (present(indep)) then
+      call read_string_1D_explicit(group_id, dims, name, buffer, indep)
+    else
+      call read_string_1D_explicit(group_id, dims, name, buffer)
+    end if
+  end subroutine read_string_1D
+
+  subroutine read_string_1D_explicit(group_id, dims, name, buffer, indep)
+    integer(HID_T), intent(in) :: group_id
+    integer(HSIZE_T), intent(in) :: dims(1)
+    character(*), intent(in)            :: name
+    character(*), intent(inout), target :: buffer(dims(1))
+    logical,      intent(in), optional  :: indep ! independent I/O
+
+    integer :: hdf5_err
+    integer :: data_xfer_mode
+#ifdef PHDF5
+    integer(HID_T) :: plist   ! property list
+#endif
+    integer(HID_T) :: dset    ! data set handle
+    integer(HID_T) :: dspace  ! data or file space handle
+    integer(HID_T) :: filetype
+    integer(HID_T) :: memtype
+    integer(HSIZE_T) :: size
+    integer(HSIZE_T) :: n
+    type(c_ptr) :: f_ptr
+
+    ! Set up collective vs. independent I/O
+    data_xfer_mode = H5FD_MPIO_COLLECTIVE_F
+    if (present(indep)) then
+      if (indep) data_xfer_mode = H5FD_MPIO_INDEPENDENT_F
+    end if
+
+    ! Get dataset and dataspace
     call h5dopen_f(group_id, trim(name), dset, hdf5_err)
     call h5dget_space_f(dset, dspace, hdf5_err)
 
+    ! Make sure buffer is large enough
+    call h5dget_type_f(dset, filetype, hdf5_err)
+    call h5tget_size_f(filetype, size, hdf5_err)
+    if (size > len(buffer(1)) + 1) then
+      call fatal_error("Character buffer is not long enough to &
+           &read HDF5 string array.")
+    end if
+
+    ! Get datatype in memory based on Fortran character
+    n = len(buffer(1))
+    call h5tcopy_f(H5T_FORTRAN_S1, memtype, hdf5_err)
+    call h5tset_size_f(memtype, n, hdf5_err)
+
+    ! Get pointer to start of string
+    f_ptr = c_loc(buffer(1)(1:1))
+
     if (using_mpio_device(group_id)) then
 #ifdef PHDF5
       call h5pcreate_f(H5P_DATASET_XFER_F, plist, hdf5_err)
       call h5pset_dxpl_mpio_f(plist, data_xfer_mode, hdf5_err)
-      call h5dread_vl_f(dset, H5T_STRING, str_tmp, dims2, dims1, hdf5_err, &
-           mem_space_id=dspace, xfer_prp=plist)
+      call h5dread_f(dset, memtype, f_ptr, hdf5_err, mem_space_id=dspace, &
```

Comment:
Ditto here and elsewhere, I will stop commenting since there may be a reason for this.